File: server/src/constants/rules.ts
export const RULE_MESSAGES = {
    FUNCTION_TOO_SHORT: "Function name \"{name}\" is too short and must be more descriptive.",
    FUNCTION_NAME_NO_VERB: "Function \"{name}\" does not start with a recognized verb.",
    FUNCTION_MIN_TWO_WORDS: "Function \"{name}\" must contain at least two words.",
    FUNCTION_TOO_LONG: "Function \"{name}\" exceeds the maximum length of {limit} lines.",
    NAME_TOO_SHORT: "\"{name}\" is too short and must be more descriptive.",
    BOOLEAN_NO_PREFIX: "Boolean variable \"{name}\" does not start with a conventional prefix.",
    BOOLEAN_NEGATIVE_PATTERN: "Boolean variable \"{name}\" has a negative naming pattern, which contradicts the positive naming convention.",
    OBJECT_KEY_TOO_SHORT: "Object key \"{name}\" is too short and must be more descriptive.",
    OBJECT_KEY_BOOLEAN_NO_PREFIX: "Object key \"{name}\" does not start with a conventional boolean prefix.",
    OBJECT_KEY_BOOLEAN_NEGATIVE_PATTERN: "Object key \"{name}\" has a negative naming pattern, which contradicts the positive naming convention.",
    CELERY_TASK_MISSING_DECORATORS: "Celery task \"{name}\" is missing required decorator(s): {decorators}.",
	CELERY_TASK_MISSING_CALLS: "Celery task \"{name}\" is missing required function call(s): {calls}.",
    THEME_SYSTEM_VIOLATION_HEXCODES: "Found a hardcoded hexcode value \"{value}\". Please use the team's theme system for colors and spacing.",
    URL_PATTERN_HARDCODED: "Avoid using hardcoded URLs in URL patterns",
    URL_PATTERN_UNNAMED: "Use named URL patterns for better reversibility and readability (e.g. 'name=user-detail')",
    URL_PATTERN_REGEX: "Prefer 'path()' over 're_path()' or 'url()' when possible",
    URL_PATTERN_INVALID_PARAMETER: "Use angle brackets '<>' for URL parameters",
    URL_PATTERN_INCONSISTENT_TRAILING_SLASH: "Use consistent trailing slashes in URL patterns",
    URL_PATTERN_MISSING_VIEW: "Ensure view is provided for URL pattern",
    URL_PATTERN_MISSING_ROUTE: "URL pattern is missing a route",
};


export enum RuleCodes {
    // Security-related rules (SEC)
    DEBUG_TRUE = "SEC01",
    HARDCODED_SECRET_KEY = "SEC02",
    EMPTY_ALLOWED_HOSTS = "SEC03",
    WILDCARD_ALLOWED_HOSTS = "SEC04",
    CSRF_COOKIE_SECURE_FALSE = "SEC05",
    SESSION_COOKIE_SECURE_FALSE = "SEC06",
    SECURE_SSL_REDIRECT_FALSE = "SEC07",
    X_FRAME_OPTIONS_NOT_SET = "SEC08",
    X_FRAME_OPTIONS_MISSING_MIDDLEWARE = "SEC09",
    SECURE_HSTS_SECONDS_NOT_SET = "SEC10",
    SECURE_HSTS_INCLUDE_SUBDOMAINS_IGNORED = "SEC11",
    SECURE_HSTS_INCLUDE_SUBDOMAINS_FALSE = "SEC12",
    RAW_SQL_USAGE = "SEC13",
    RAW_SQL_USAGE_WITH_CURSOR = "SEC14",

    // Complexity-related rules (CMP)
    COMPLEX_VIEW = "CMP01",
    FUNCTION_COMPLEXITY = "CMP02",

    // Code Quality-related rules (CDQ)
    NO_EXCEPTION_HANDLER = "CDQ01",
    VARIABLE_NAME_TOO_SHORT = "CDQ02",
    OBJECT_PROPERTY_NAME_TOO_SHORT = "CDQ03",
    FUNCTION_NAME_TOO_SHORT = "CDQ04",
    FUNCTION_NAME_NO_VERB = "CDQ05",
    FUNCTION_TOO_LONG = "CDQ06",
    CLASS_NAME_CONVENTION = "CDQ07",
    LIST_NAME_CONVENTION = "CDQ08",
    DICTIONARY_VALIDATION = "CDQ09",
    FOR_LOOP_TARGET_VALIDATION = "CDQ010",
    DJANGO_MODEL_FIELD_NAMING = "CDQ011",
    DJANGO_SERIALIZER_FIELD_NAMING = "CDQ12",
    DJANGO_FIELD_CONVENTIONS = "CDQ13",
    COMMENT_VALIDATION = "CDQ14",
    CELERY_TASK_VALIDATION = "CDQ16",

    // Style-related rules (STY)
    BOOLEAN_VARIABLE_PREFIX = "STY01",
    BOOLEAN_VARIABLE_POSITIVE_NAMING = "STY02",
    BOOLEAN_PROPERTY_PREFIX = "STY03",
    BOOLEAN_PROPERTY_POSITIVE_NAMING = "STY04",

    // Configuration-related rules (CFG)
    RESERVED_SYMBOL_HANDLING = "CFG01",

    // Performance-related rules (PER)
    NPLUSONE = "PER01",
}

File: server/bundled/tools/python/djangoly/core/lib/rules.py
from typing import Dict
from enum import Enum

class RuleCode(Enum):
    # Security-related Rules (SEC)
    SEC01 = "SEC01"
    SEC02 = "SEC02"
    SEC03 = "SEC03"
    SEC04 = "SEC04"
    SEC05 = "SEC05"
    SEC06 = "SEC06"
    SEC07 = "SEC07"
    SEC08 = "SEC08"
    SEC09 = "SEC09"
    SEC10 = "SEC10"
    SEC11 = "SEC11"
    SEC12 = "SEC12"
    SEC13 = "SEC13"
    SEC14 = "SEC14"

    # Code Quality-related Rules (CDQ)
    CDQ01 = "CDQ01"
    CDQ02 = "CDQ02"
    CDQ03 = "CDQ03"
    CDQ04 = "CDQ04"
    CDQ05 = "CDQ05"
    CDQ06 = "CDQ06"
    CDQ07 = "CDQ07"

    # Style-related Rules (STY)
    STY01 = "STY01"
    STY02 = "STY02"
    STY03 = "STY03"


class Rule:
    def __init__(self, code: RuleCode, name: str, description: str, additional_info: str = None):
        self.code = code
        self.name = name
        self.description = description
        self.additional_info = additional_info


RULES: Dict[RuleCode, Rule] = {
    # Security-related Rules (SEC)
    RuleCode.SEC01: Rule(
        code=RuleCode.SEC01,
        name="DEBUG_TRUE",
        description="DEBUG is set to True. Ensure it is False in production."
    ),
    RuleCode.SEC02: Rule(
        code=RuleCode.SEC02,
        name="HARDCODED_SECRET_KEY",
        description="SECRET_KEY appears to be hardcoded. It is strongly recommended to store it in an environment variable."
    ),
    RuleCode.SEC03: Rule(
        code=RuleCode.SEC03,
        name="EMPTY_ALLOWED_HOSTS",
        description="ALLOWED_HOSTS is empty. This is not secure for production."
    ),
    RuleCode.SEC04: Rule(
        code=RuleCode.SEC04,
        name="WILDCARD_ALLOWED_HOSTS",
        description="ALLOWED_HOSTS contains a wildcard '*'. This is not recommended for production."
    ),
    RuleCode.SEC05: Rule(
        code=RuleCode.SEC05,
        name="CSRF_COOKIE_SECURE_FALSE",
        description="CSRF_COOKIE_SECURE is False. Set this to True to avoid transmitting the CSRF cookie over HTTP accidentally."
    ),
    RuleCode.SEC06: Rule(
        code=RuleCode.SEC06,
        name="SESSION_COOKIE_SECURE_FALSE",
        description="SESSION_COOKIE_SECURE is False. Set this to True to avoid transmitting the session cookie over HTTP accidentally."
    ),
    RuleCode.SEC07: Rule(
        code=RuleCode.SEC07,
        name="SECURE_SSL_REDIRECT_FALSE",
        description="SECURE_SSL_REDIRECT is set to False. It should be True in production to enforce HTTPS."
    ),
    RuleCode.SEC08: Rule(
        code=RuleCode.SEC08,
        name="X_FRAME_OPTIONS_NOT_SET",
        description="X_FRAME_OPTIONS is not set to a valid value. It should be either 'DENY' or 'SAMEORIGIN' to prevent clickjacking."
    ),
    RuleCode.SEC09: Rule(
        code=RuleCode.SEC09,
        name="X_FRAME_OPTIONS_MISSING_MIDDLEWARE",
        description="X_FRAME_OPTIONS is set, but the XFrameOptionsMiddleware is missing from the MIDDLEWARE list."
    ),
    RuleCode.SEC10: Rule(
        code=RuleCode.SEC10,
        name="SECURE_HSTS_SECONDS_NOT_SET",
        description="SECURE_HSTS_SECONDS is set to 0. Set it to a positive value to enforce HTTPS."
    ),
    RuleCode.SEC11: Rule(
        code=RuleCode.SEC11,
        name="SECURE_HSTS_INCLUDE_SUBDOMAINS_IGNORED",
        description="SECURE_HSTS_INCLUDE_SUBDOMAINS is True, but it has no effect because SECURE_HSTS_SECONDS is 0."
    ),
    RuleCode.SEC12: Rule(
        code=RuleCode.SEC12,
        name="SECURE_HSTS_INCLUDE_SUBDOMAINS_FALSE",
        description="SECURE_HSTS_INCLUDE_SUBDOMAINS is set to False. Set it to True for better security."
    ),
    RuleCode.SEC13: Rule(
        code=RuleCode.SEC13,
        name="RAW_SQL_USAGE",
        description="Avoid using 'raw' queries to execute SQL directly, bypassing Django's ORM protections."
    ),
    RuleCode.SEC14: Rule(
        code=RuleCode.SEC14,
        name="RAW_SQL_USAGE_WITH_CURSOR",
        description="Avoid using 'connection.cursor()' to execute SQL directly, bypassing Django's ORM protections."
    ),

    # Code Quality-related Rules (CDQ)
    RuleCode.CDQ01: Rule(
        code=RuleCode.CDQ01,
        name="NO_EXCEPTION_HANDLER",
        description="Checks for the presence of exception handlers in views."
    ),
    RuleCode.CDQ02: Rule(
        code=RuleCode.CDQ02,
        name="NAME_TOO_SHORT",
        description="Names (variables, functions, object properties) should be at least 3 characters long (excluding leading underscores).",
        additional_info="Applies to all variables, function names, object properties, etc. Excludes 'id' for object properties."
    ),
    RuleCode.CDQ03: Rule(
        code=RuleCode.CDQ03,
        name="FUNCTION_NAME_NO_VERB",
        description="Function names should include a verb to describe the action.",
        additional_info="Applies to regular functions, Django model methods, serializer methods, view methods, and testcase methods."
    ),
    RuleCode.CDQ04: Rule(
        code=RuleCode.CDQ04,
        name="FUNCTION_TOO_LONG",
        description="Functions should not exceed a specified number of lines.",
        additional_info="The maximum allowed length is configurable."
    ),
    RuleCode.CDQ05: Rule(
        code=RuleCode.CDQ05,
        name="DJANGO_FIELD_CONVENTIONS",
        description="Django fields should follow specific conventions."
    ),
    RuleCode.CDQ06: Rule(
        code=RuleCode.CDQ06,
        name="REDUNDANT_QUERY_METHODS",
        description="Identifies redundant QuerySet method chains, such as `all().filter()` or `filter().all()`. Simplified queries avoid redundant operations."
    ),
    RuleCode.CDQ07: Rule(
        code=RuleCode.CDQ07,
        name="SKINNY_VIEWS",
        description="Views should be skinny and delegate business logic to services or model methods."
    ),

    # Style-related Rules (STY)
    RuleCode.STY01: Rule(
        code=RuleCode.STY01,
        name="BOOLEAN_VARIABLE_PREFIX",
        description="Boolean variables should use required prefixes.",
        additional_info="Prefixes are configurable in settings."
    ),
    RuleCode.STY02: Rule(
        code=RuleCode.STY02,
        name="BOOLEAN_VARIABLE_POSITIVE_NAMING",
        description="Boolean variables should use positive naming (avoid negative patterns)."
    ),
    RuleCode.STY03: Rule(
        code=RuleCode.STY03,
        name="TEST_NAMING_CONVENTION",
        description="Test names should follow the pattern `test_{result}_given_{conditions}`."
    ),
}

